lib/services/label_expressions.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package services
+
+import (
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/gravitational/trace"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/sirupsen/logrus"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/typical"
+)
+
+type labelExpression typical.Expression[labelExpressionEnv, bool]
+
+type labelExpressionEnv struct {
+	resourceLabelGetter LabelGetter
+	userTraits          map[string][]string
+}
+
+func parseLabelExpression(expr string) (labelExpression, error) {
+	if parsedExpr, ok := labelExpressionCache.Get(expr); ok {
+		return parsedExpr, nil
+	}
+	parsedExpr, err := labelExpressionParser.Parse(expr)
+	if err != nil {
+		return nil, trace.Wrap(err, "parsing label expression")
+	}
+	if evicted := labelExpressionCache.Add(expr, parsedExpr); evicted {
+		logrus.Warn("evicting entry from label expression cache")
+	}
+	return parsedExpr, nil
+
+}
+
+var (
+	labelExpressionCache  = mustNewLabelExpressionCache()
```

Comment:
What is the reasoning behind having a label expressions cache?

Is the parsing expensive?

Which sources of expressions do we expect the cache to cover?

Reply:
https://github.com/gravitational/teleport/blob/master/rfd/0116-label-expressions.md#performance

Parsing is usually quite expensive compared to evaluation. My main performance concern is operations like `ListNodes` where we check access to each and every node in the cluster. If you have a role with a simple expression like `labels["env"] == "staging"`, and you call `ListNodes` with 50k nodes, without a cache we would be parsing that expression 50k times. With a cache we only parse it once, after that it's a simple string comparison.

I ran a couple benchmarks without the cache and ListNodes got like 2-3x slower. With the cache, `node_labels_expression: 'labels["env"] == staging'` looks like it's almost 30% faster than
```
node_labels:
  env: staging
```

This cache will cover all "label expressions", which will show up in role specs, `node_labels_expression`, `app_labels_expression`, etc...

Reply:
Gotcha, thanks for the explanation.